filesystems/memory.py

```diff
@@ -374,8 +375,14 @@ def FS(self, name):
             readlink=_fs(self.readlink),
         )()
 
-    def create_directory(self, path):
-        self[path].create_directory(path=path)
+    def create_directory(self, path, parents=False):
+        if parents:
+            for maybe_parent in path.heritage():
+                a = self[maybe_parent]
+                if isinstance(a, _DirectoryChild):
+                    a.create_directory(path=maybe_parent)
```

Comment:
@Julian, this is all around pretty gross.  `maybe_parent`?  `a`?  `isinstance()`?  There's no file system here afaik so we can't `.is_dir()`.  I'm not sure what to call each element of `.heritage()` since `ancestor` lies about the final yielded value (`path` itself).

Reply:
Usually in those circumstances I just call the element `each` -- but here yeah to avoid all of this is why there are distinct classes -- basically the logic can go up a level.

So instead of needing to check whether you have a `_DirectoryChild`, that logic should go inside each implementation. E.g., like 2 implementations should do nothing -- `_File`, because it knows it already exists, `_Directory`, same, etc.

Can write this out a bit more clearly (maybe in a second, first diagnosing some unrelated build error).

Reply:
But tl;dr yeah you don't have `is_dir`, in exchange you just literally know which type of thing you are based on which class you're implementing.

Reply:
I made a change for this... but I'm a bit confused by the `create_directory()` implementations.  They take a path but also know that the path is exactly themselves and so an `_Directory` can just fail to create regardless of what path is requested?

Reply:
It doesn't have to be the same `path` -- e.g., if the path was a resolved symlink.

Essentially (up until now?) the argument exists effectively to raise the right original path in exceptions.

Reply:
I.e., a `_File` object doesn't have to know what path it lives at, same as how on an abstract filesystem the same file can live in multiple places.